action/governance/cancelProposal.go

```diff
@@ -0,0 +1,166 @@
+	}
+
+	err = action.ValidateFee(ctx.FeePool.GetOpt(), tx.Fee)
+	if err != nil {
+		return false, err
+	}
+
+	// validate params
+	if err = cc.ProposalId.Err(); err != nil {
+		return false, errors.Wrap(err, "invalid proposal id")
+	}
+	if err = cc.Proposer.Err(); err != nil {
+		return false, errors.Wrap(err, "invalid proposer address")
+	}
+
+	return true, nil
+}
+
+func (c cancelProposalTx) ProcessCheck(ctx *action.Context, tx action.RawTx) (bool, action.Response) {
+	ctx.Logger.Debug("ProcessCheck CancelProposalTx transaction for CheckTx", tx)
+	return runCancel(ctx, tx)
+}
+
+func (c cancelProposalTx) ProcessFee(ctx *action.Context, signedTx action.SignedTx, start action.Gas, size action.Gas) (bool, action.Response) {
+	return action.BasicFeeHandling(ctx, signedTx, start, size, 2)
+}
+
+func (c cancelProposalTx) ProcessDeliver(ctx *action.Context, tx action.RawTx) (bool, action.Response) {
+	ctx.Logger.Debug("ProcessDeliver CancelProposalTx transaction for DeliverTx", tx)
+	return runCancel(ctx, tx)
+}
+
+func runCancel(ctx *action.Context, tx action.RawTx) (bool, action.Response) {
+	cc := &CancelProposal{}
+	err := cc.Unmarshal(tx.Data)
+	if err != nil {
+		return false, action.Response{Log: "calcel proposal failed, deserialization err"}
+	}
+
+	// Get Proposal from proposal ACTIVE store
+	pms := ctx.ProposalMasterStore
+	proposal, err := pms.Proposal.WithPrefixType(gov.ProposalStateActive).Get(cc.ProposalId)
+	if err != nil {
+		return false, action.Response{
+			Log: fmt.Sprintf("calcel proposal failed, id= %v, proposal not found in ACTIVE store", cc.ProposalId)}
+	}
+
+	// Check if proposal is in FUNDING status
+	if proposal.Status != gov.ProposalStatusFunding {
+		return false, action.Response{
+			Log: fmt.Sprintf("cancel proposal failed, id= %v, proposal not in FUNDING status", cc.ProposalId)}
+	}
+
+	// Check if proposal funding height is passed
+	if ctx.Header.Height > proposal.FundingDeadline {
+		return false, action.Response{
+			Log: fmt.Sprintf("cancel proposal failed, id= %v, funding height passed", cc.ProposalId)}
+	}
+
+	// Check if proposer matches
+	if !proposal.Proposer.Equal(cc.Proposer) {
+		return false, action.Response{
+			Log: fmt.Sprintf("cancel proposal failed, id= %v, proposer does not match", cc.ProposalId)}
+	}
+
+	// Update fields and add it to FAILED store
+	proposal.Status = gov.ProposalStatusCompleted
+	proposal.Outcome = gov.ProposalOutcomeCancelled
+	proposal.Description += " - Canceled"
+	if cc.Reason != "" {
+		proposal.Description += fmt.Sprintf(" for reason: %v", cc.Reason)
+	}
+	err = pms.Proposal.WithPrefixType(gov.ProposalStateFailed).Set(proposal)
+	if err != nil {
+		return false, action.Response{
+			Log: fmt.Sprintf("cancel proposal failed, id= %v, failed to add proposal to FAILED store", cc.ProposalId)}
+	}
+
+	// Delete proposal in ACTIVE store
+	ok, err := pms.Proposal.WithPrefixType(gov.ProposalStateActive).Delete(cc.ProposalId)
+	if err != nil || !ok {
+		return false, action.Response{
```

Comment:
Might have to panic if it failed to delete.

Reply:
Yes. You're right. I wanted to add this panic until I saw that the Delete always returns a Nil.